Source/Classes/AnimatedImages/PINAnimatedImage.m

```diff
@@ -0,0 +1,101 @@
+//
+//  PINAnimatedImage.m
+//  PINRemoteImage
+//
+//  Created by Garrett Moon on 9/17/17.
+//  Copyright © 2017 Pinterest. All rights reserved.
+//
+
+#import "PINAnimatedImage.h"
+
+NSString *kPINAnimatedImageErrorDomain = @"kPINAnimatedImageErrorDomain";
+
+const NSTimeInterval kPINAnimatedImageDisplayRefreshRate = 60.0;
+//http://nullsleep.tumblr.com/post/16524517190/animated-gif-minimum-frame-delay-browser
+const Float32 kPINAnimatedImageMinimumDuration = 1 / kPINAnimatedImageDisplayRefreshRate;
+const Float32 kPINAnimatedImageDefaultDuration = 0.1;
+
+@interface PINAnimatedImage ()
+{
+    CFTimeInterval _totalDuration;
+}
+@end
+
+@implementation PINAnimatedImage
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        _totalDuration = -1;
+    }
+    return self;
+}
+
+- (CFTimeInterval)durationAtIndex:(NSUInteger)index
+{
+    NSAssert(NO, @"Must be overridden by subclass");
+    return 0;
+}
+
+- (size_t)frameCount
+{
+    NSAssert(NO, @"Must be overridden by subclass");
+    return 0;
+}
+
+
+- (CFTimeInterval)totalDuration
+{
+    if (_totalDuration == -1) {
+        _totalDuration = 0;
+        for (NSUInteger idx = 0; idx < self.frameCount; idx++) {
+            _totalDuration += [self durationAtIndex:idx];
+        }
+    }
+
+    return _totalDuration;
+}
+
+- (NSUInteger)frameInterval
+{
+    return MAX(self.minimumFrameInterval * kPINAnimatedImageDisplayRefreshRate, 1);
+}
+
+//Credit to FLAnimatedImage ( https://github.com/Flipboard/FLAnimatedImage ) for display link interval calculations
```

Comment:
Not necessary at the moment, but might be good to unit test this piece. Always seemed a bit risky to me!